src/code/z_play.c

```diff
@@ -1119,168 +1119,170 @@ void Play_Draw(PlayState* this) {
-                    Gfx* gfxP = OVERLAY_DISP;
+        if ((R_HREG_MODE != HREG_MODE_PLAY) || R_PLAY_DRAW_SANDSTORM) {
+            if (this->envCtx.sandstormState != SANDSTORM_OFF) {
+                Environment_DrawSandstorm(this, this->envCtx.sandstormState);
+            }
+        }
 
-                    // Copy the frame buffer contents at this point in the display list to the zbuffer
-                    // The zbuffer must then stay untouched until unpausing
-                    this->pauseBgPreRender.fbuf = gfxCtx->curFrameBuffer;
-                    this->pauseBgPreRender.fbufSave = (u16*)gZBuffer;
-                    PreRender_SaveFramebuffer(&this->pauseBgPreRender, &gfxP);
-                    if (R_PAUSE_BG_PRERENDER_STATE == PAUSE_BG_PRERENDER_SETUP) {
-                        this->pauseBgPreRender.cvgSave = (u8*)gfxCtx->curFrameBuffer;
-                        PreRender_DrawCoverage(&this->pauseBgPreRender, &gfxP);
+        if ((R_HREG_MODE != HREG_MODE_PLAY) || R_PLAY_DRAW_DEBUG_OBJECTS) {
+            DebugDisplay_DrawObjects(this);
+        }
 
-                        R_PAUSE_BG_PRERENDER_STATE = PAUSE_BG_PRERENDER_PROCESS;
-                    } else {
-                        gTransitionTileState = TRANS_TILE_PROCESS;
-                    }
-                    OVERLAY_DISP = gfxP;
-                    this->unk_121C7 = 2;
-                    R_GRAPH_TASKSET00_FLAGS |= 1;
-                } else {
-                Play_Draw_DrawOverlayElements:
-                    if ((R_HREG_MODE != HREG_MODE_PLAY) || R_PLAY_DRAW_OVERLAY_ELEMENTS) {
-                        Play_DrawOverlayElements(this);
-                    }
-                }
+        if ((R_PAUSE_BG_PRERENDER_STATE == PAUSE_BG_PRERENDER_SETUP) || (gTransitionTileState == TRANS_TILE_SETUP)) {
+            Gfx* gfxP = OVERLAY_DISP;
+
+            // Copy the frame buffer contents at this point in the display list to the zbuffer
+            // The zbuffer must then stay untouched until unpausing
+            this->pauseBgPreRender.fbuf = gfxCtx->curFrameBuffer;
+            this->pauseBgPreRender.fbufSave = (u16*)gZBuffer;
+            PreRender_SaveFramebuffer(&this->pauseBgPreRender, &gfxP);
+            if (R_PAUSE_BG_PRERENDER_STATE == PAUSE_BG_PRERENDER_SETUP) {
+                this->pauseBgPreRender.cvgSave = (u8*)gfxCtx->curFrameBuffer;
+                PreRender_DrawCoverage(&this->pauseBgPreRender, &gfxP);
+
+                R_PAUSE_BG_PRERENDER_STATE = PAUSE_BG_PRERENDER_PROCESS;
+            } else {
+                gTransitionTileState = TRANS_TILE_PROCESS;
             }
+            OVERLAY_DISP = gfxP;
+            this->unk_121C7 = 2;
+            R_GRAPH_TASKSET00_FLAGS |= 1;
+            goto Play_Draw_skip;
+        }
+
+    Play_Draw_DrawOverlayElements:
+        if ((R_HREG_MODE != HREG_MODE_PLAY) || R_PLAY_DRAW_OVERLAY_ELEMENTS) {
+            Play_DrawOverlayElements(this);
         }
     }
 
+Play_Draw_skip:
```

Comment:
```suggestion
Play_Draw_Skip:
```
Not sure we have a particular style for labels, but the other one is capitalized, so seems appropriate for this one to be as well.

Reply:
Worth noting this extra `Play_Draw_Skip` goto introduced here is actually required in MM's `Play_DrawMain` to fix regalloc. I wonder if it would also be required in OoT's retail where these go-to signals (not sure what to call it, usually see them as `dummy:;` in MM) does affect regalloc more.

Reply:
I might have a macro or two off for retail, but it does look this this goto will very likely be required for OoT retail, at least looking at 1.0's `Play_Draw`

https://decomp.me/scratch/bCuGH